prov/rxd/src/rxd_av.c

```diff
@@ -103,68 +86,39 @@ static int rxd_av_set_addrlen(struct rxd_av *av, const void *addr)
 	return ret;
 }
 
-static fi_addr_t rxd_av_dg_addr(struct rxd_av *av, fi_addr_t fi_addr)
+int rxd_av_insert_fi_addr(struct rxd_av *av, fi_addr_t *dg_addr, 
+                    fi_addr_t *fi_addr, uint64_t flags)
 {
-	fi_addr_t rxd_addr = av->fi_addr_table[fi_addr];
-
-	return rxd_addr == FI_ADDR_UNSPEC ? rxd_addr :
-		av->rxd_addr_table[rxd_addr].dg_addr;
-}
-
-static fi_addr_t rxd_set_rxd_addr(struct rxd_av *av, fi_addr_t dg_addr)
-{
-	int tries = 0;
-
-	while (av->rxd_addr_table[av->rxd_addr_idx].dg_addr != FI_ADDR_UNSPEC &&
-	       tries < av->util_av.count) {
-		if (++av->rxd_addr_idx == av->util_av.count)
-			av->rxd_addr_idx = 0;
-		tries++;
-	}
-	assert(av->rxd_addr_idx < av->util_av.count && tries < av->util_av.count);
-	av->rxd_addr_table[av->rxd_addr_idx].dg_addr = dg_addr;
-
-	return av->rxd_addr_idx;
-}
-
-static fi_addr_t rxd_set_fi_addr(struct rxd_av *av, fi_addr_t rxd_addr)
-{
-	int tries = 0;
-
-	while (av->fi_addr_table[av->fi_addr_idx] != FI_ADDR_UNSPEC &&
-	       tries < av->util_av.count) {
-		if (++av->fi_addr_idx == av->util_av.count)
-			av->fi_addr_idx = 0;
-		tries++;
-	}
-	assert(av->fi_addr_idx < av->util_av.count && tries < av->util_av.count);
-	av->fi_addr_table[av->fi_addr_idx] = rxd_addr;
-	av->rxd_addr_table[rxd_addr].fi_addr = av->fi_addr_idx;
-
-	return av->fi_addr_idx;
+    int ret; 
+    ret = ofi_av_insert_addr(&av->util_av, dg_addr, fi_addr);
+   	if (ret) {
+       	FI_WARN(&rxd_prov, FI_LOG_AV,
+			"failed to insert address: %d (%s)\n", 
+            -ret, fi_strerror(-ret));
+        fi_av_remove(av->dg_av, dg_addr, 1, flags);
+        return -FI_EINVAL;
+    }
+    return ret;
 }
 
 int rxd_av_insert_dg_addr(struct rxd_av *av, const void *addr,
-			  fi_addr_t *rxd_addr, uint64_t flags,
+			  fi_addr_t *dg_addr, uint64_t flags,
 			  void *context)
-{
-	fi_addr_t dg_addr;
+{	
 	int ret;
+    size_t len;
+    len = RXD_NAME_LENGTH;
+    uint8_t tmp_addr[RXD_NAME_LENGTH];
 
-	ret = fi_av_insert(av->dg_av, addr, 1, &dg_addr,
+    ret = fi_av_lookup(av->dg_av, *dg_addr, tmp_addr, &len);
+    if (ret) {
+    	ret = fi_av_insert(av->dg_av, addr, 1, dg_addr,
 			     flags, context);
-	if (ret != 1)
-		return -FI_EINVAL;
-
-	*rxd_addr = rxd_set_rxd_addr(av, dg_addr);
-
-	ret = ofi_rbmap_insert(&av->rbmap, (void *) addr, (void *) (*rxd_addr),
-			       NULL);
-	if (ret) {
-		assert(ret != -FI_EALREADY);
-		fi_av_remove(av->dg_av, &dg_addr, 1, flags);
-	}
-
+	    if (ret != 1){
+            FI_WARN(&rxd_prov, FI_LOG_AV, "failed to insert in dg AV");
+		    return -FI_EINVAL;
+        }
+    }
```

Comment:
I would outdent this and return earlier if it's already there. ie
```
ret = fi_av_lookup();
if (!ret)
    return FI_SUCCESS;
ret = fi_av_insert();
...
```

Reply:
I have changed this now to use a hashtable to check if a mapping for addr-->dgaddr exists or not. fi_av_lookup was incorrectly used in this code. Now the function rxd_dg_av_insert will always return a valid dg_addr.